File: kafkakewl-deploy/src/main/scala/com/mwam/kafkakewl/deploy/services/TopologyDeploymentsService.scala
/*
 * SPDX-FileCopyrightText: 2023 Marshall Wace <opensource@mwam.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.mwam.kafkakewl.deploy.services

import com.mwam.kafkakewl.common.persistence.PersistentStore
import com.mwam.kafkakewl.domain.*
import com.mwam.kafkakewl.domain.validation.*
import zio.*

class TopologyDeploymentsService private (
    private val persistentStore: PersistentStore,
    private val topologyDeploymentsToKafkaService: TopologyDeploymentsToKafkaService,
    private val mutex: Semaphore,
    private val topologyDeploymentsRef: Ref[TopologyDeployments]
) {
  def deploy(deployments: Deployments): IO[PostDeploymentsFailure, DeploymentsSuccess] =
    mutex.withPermit {
      for {
        _ <- ZIO.logInfo(s"deploying $deployments")

        // TODO authorization

        // Validation before deployment
        topologyDeploymentsBefore <- topologyDeploymentsRef.get
        _ <- DeploymentsValidation
          .validate(topologyDeploymentsBefore.toTopologies, deployments)
          .toZIOParallelErrors
          .mapError(DeploymentsFailure.validation)

        // Performing the actual deployment to kafka (dryRun is respected here)
        success <- topologyDeploymentsToKafkaService
          .deploy(deployments)
          .logError("deploying TopologyDeployments")
          .mapError(DeploymentsFailure.deployment)

        // Creating the new TopologyDeployments from the statuses
        topologyDeployments = deployments.deploy
          .map(t => (t.id, TopologyDeployment(t.id, success.statuses(t.id), Some(t))))
          .toMap ++ deployments.delete
          .map(tid => (tid, TopologyDeployment(tid, success.statuses(tid), None)))
          .toMap

        _ <- (for {
          // Persisting the new TopologyDeployments
          _ <- persistentStore.save(topologyDeployments).logError("saving TopologyDeployments").mapError(DeploymentsFailure.persistence)

          // Updating the in-memory state if everything succeeded so far
          _ <- topologyDeploymentsRef.update {
            _ ++ topologyDeployments -- deployments.delete
          }
        } yield ()).unless(deployments.options.dryRun) // Not making any actual changes if dryRun = true

        _ <- ZIO.logInfo(s"finished deploying $deployments")
      } yield DeploymentsSuccess(
        topologyDeployments
          .map((tid, td) => (tid, td.status))
          .toMap
      )
    }

  def getTopologyDeployments(topologyDeploymentQuery: TopologyDeploymentQuery): UIO[Seq[TopologyDeployment]] = for {
    // TODO use the query
    topologyDeployments <- topologyDeploymentsRef.get
  } yield topologyDeployments.values.toList

  def getTopologyDeployment(id: TopologyId): UIO[Option[TopologyDeployment]] = for {
    topologies <- topologyDeploymentsRef.get
  } yield topologies.get(id)
}

object TopologyDeploymentsService {
  def live: ZLayer[PersistentStore & TopologyDeploymentsToKafkaService, Nothing, TopologyDeploymentsService] =
    ZLayer.fromZIO {
      for {
        persistentStore <- ZIO.service[PersistentStore]
        topologyDeploymentsToKafkaService <- ZIO.service[TopologyDeploymentsToKafkaService]
        // TODO what happens if we fail here?
        topologyDeployments <- persistentStore.loadLatest().orDie
        mutex <- Semaphore.make(permits = 1)
        topologyDeploymentsRef <- Ref.make(topologyDeployments)
      } yield TopologyDeploymentsService(persistentStore, topologyDeploymentsToKafkaService, mutex, topologyDeploymentsRef)
    }
}

File: kafkakewl-deploy/src/main/kotlin/com/mwam/kafkakewl/deploy/services/TopologyDeploymentsService.kt
/*
 * SPDX-FileCopyrightText: 2024 Marshall Wace <opensource@mwam.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.mwam.kafkakewl.deploy.services

import arrow.core.Either
import arrow.core.right
import com.mwam.kafkakewl.common.persistence.PersistentStore
import com.mwam.kafkakewl.domain.Deployments
import com.mwam.kafkakewl.domain.DeploymentsSuccess
import com.mwam.kafkakewl.domain.PostDeploymentsFailure
import com.mwam.kafkakewl.domain.TopologyDeployment
import com.mwam.kafkakewl.domain.TopologyDeploymentQuery
import com.mwam.kafkakewl.domain.TopologyDeploymentStatus
import com.mwam.kafkakewl.domain.TopologyDeployments
import com.mwam.kafkakewl.domain.TopologyId
import io.github.oshai.kotlinlogging.KotlinLogging
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.concurrent.atomic.AtomicReference

interface TopologyDeploymentsService {
    suspend fun deploy(deployments: Deployments): Either<PostDeploymentsFailure, DeploymentsSuccess>
    fun getTopologyDeployments(topologyDeploymentQuery: TopologyDeploymentQuery): List<TopologyDeployment>
    fun getTopologyDeployment(id: TopologyId): TopologyDeployment?
}

private val logger = KotlinLogging.logger {}

class TopologyDeploymentsServiceImpl private constructor(
    private val persistentStore: PersistentStore,
    private val topologyDeploymentsRef: AtomicReference<TopologyDeployments>
) : TopologyDeploymentsService {
    companion object {
        fun create(persistentStore: PersistentStore): TopologyDeploymentsServiceImpl {
            // TODO is there a better way? Maybe loadLatest() could just throw instead of returning a Result<TopologyDeployment>?
            val topologyDeployments = persistentStore.loadLatest().getOrThrow()
            logger.info { "Loaded ${topologyDeployments.size} topology deployments" }
            return TopologyDeploymentsServiceImpl(persistentStore, AtomicReference(topologyDeployments))
        }
    }

    private val mutex = Mutex()

    override suspend fun deploy(deployments: Deployments): Either<PostDeploymentsFailure, DeploymentsSuccess> {
        logger.info { "Trying to enter the deploy-mutex..." }
        mutex.withLock {
            val deploymentsLog = "${deployments.deploy.size}, deleting ${deployments.delete.size} topologies with options = ${deployments.options} (deploying ${deployments.deploy.map { it.id }}, deleting ${deployments.delete})"
            logger.info { "Deploying $deploymentsLog..." }

            // TODO authorization
            // TODO Validation before deployment
            // TODO Performing the actual deployment to kafka (dryRun is respected here)
            // For now we just fake successful deployments and deletes
            val success = DeploymentsSuccess(
                (deployments.deploy.map { it.id } + deployments.delete)
                    .associate { it to TopologyDeploymentStatus() }
            )
            // TODO Creating the new TopologyDeployments from the statuses
            val topologyDeploymentsDeployed = deployments.deploy.associate { it.id to TopologyDeployment(it.id, success.statuses[it.id]!!, it) }
            val topologyDeploymentsDeleted = deployments.delete.associate { it to TopologyDeployment(it, success.statuses[it]!!, null) }
            val topologyDeploymentsDeployedDeleted = topologyDeploymentsDeployed + topologyDeploymentsDeleted

            if (!deployments.options.dryRun) {
                // Persisting the new TopologyDeployments
                persistentStore.save(topologyDeploymentsDeployedDeleted)
                // Updating the in-memory state if everything succeeded so far
                topologyDeploymentsRef.set(topologyDeploymentsRef.get() + topologyDeploymentsDeployed - topologyDeploymentsDeleted.keys)
            }

            logger.info { "Finished deploying $deploymentsLog" }

            return success.right()
        }
    }

    override fun getTopologyDeployments(query: TopologyDeploymentQuery): List<TopologyDeployment> {
        // getting the filtered, ordered list of topology deployments
        val filterRegexOrNull = query.topologyIdFilterRegex?.toRegex()
        val topologyDeployments = topologyDeploymentsRef.get().values
            .filter { filterRegexOrNull == null || filterRegexOrNull.matches(it.topologyId.value) }
            .sortedBy { it.topologyId.value }

        // selecting a range based on offset/limit
        val fromIndex = query.offset ?: 0
        val toIndex = (query.limit?.let { it + fromIndex } ?: topologyDeployments.size).coerceAtMost(topologyDeployments.size)
        return topologyDeployments
            .subList(fromIndex, toIndex)
            .toList()
    }

    override fun getTopologyDeployment(id: TopologyId): TopologyDeployment? = topologyDeploymentsRef.get()[id]
}
